<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="UTF-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
		<script src="app.js"></script>
		<title>Async JavaScript</title>
	</head>
	<body>
		<h1>Async JavaScript</h1>
		<h3>Synchronous Code</h3>
		<p>
			JavaScript is single-threaded meaning that only one task/piece of code can
			be exectured at a time due to their only being one thread available to run
			the exection. This means JavaScript code runs from top to bottom line by
			line and step by step.
		</p>
		<h3>Asynchronous Code</h3>
		<p>
			Sometimes we have certain operations in our code that takes longer such as
			timers and HTTP requests. Now we wouldnt want these lengthy operations to
			execute and block the main thread as then the rest of our synchronous code
			cannot run until the operation is complete. To deal with these lengthy
			operations we can send them off to the browser which uses multiple threads
			to deal with such as browser APIs for timers like setTimeout and
			setInterval, event listeners or HTTP requests with fetch. These browser
			APIs typically take a callback fuction that is called later on in the
			single threaded synchronous JavaScript once the lenghty operarion has
			completed.
		</p>
		<h3>Event Loop</h3>
		<p>
			The event loop is used in browsers to manage asynchronous code and
			callbacks. For example say we use a setTimeout that setTimeout will
			execute immediately as a browser API the browser will handle the setting
			of this timer and then add the callback to what is known as the message
			queue. Then after your setTimeout the rest of the synchronous code carries
			on running as normal, After the timer has finished the browser event loop
			will look for a time when the call stack is empty to then go ahead and add
			that queued message being our callback to the stack to then be executed as
			part of synchronous code. The event loop is what snchronises the message
			queue witht the call stack. The event loop is always running checking for
			if the call stack is empty and if it is adds any messages queued that need
			to be added to the stack.
		</p>
		<h3>Promises</h3>
		<p>
			Promises allow us to handle our asyncchronous code/callbacks in a more
			synchronous looking way and helps us avoid the confusion of 'callback
			hell' when we have multiple nested callbacks that can be hard to follow.
			Promises allow us to chain callback executions together with ''.then'
			following on from the previous calls return and therefore keep the code on
			one nested level instead of possible multi nested levels when there is
			multiple callbacks. Promises are an object that takes a function as an
			argument with that function containing two more functions being a resolve
			and reject function and can be created with the 'new Promise()'' syntax.
			The '.then' chaining on a promise begins after resolve has executed and a
			promise can resolve multiple times hence multiple '.then' chaining.
		</p>
		<h3>Chaining Multiple Promises</h3>
		<p>
			Promises can be chained with multiple '.then' by simply adding a return in
			the previous '.then' block. Doing this sets the promise from resolved back
			to pending as it automatically wraps a new promise on the return from the
			previous then. Your next then can then get the data from that previous
			return as a parameter in the callback of '.then' and the chain can keep
			continuing like this. Note that you dont have to return a promise to chain
			multiple '.then' anything you return gets wrapped in a promise and sets
			the promise from resolved back to pending.
		</p>
		<h3>Promise Error Handling</h3>
		<p>
			Promises take a function as an argument which in turn takes a resolve and
			a reject function as arguments. The reject function is used for error
			handling when the promise isnt fulfilled and resolved. When creating our
			promises we wrap the error state with the reject function call and pass in
			any error related data to it. Then when it comes to using the promise and
			chaining it we can handle the error we get from reject by passing a second
			argument to '.then' which is our reject error handling for that '.then'
			case or we can chain '.catch'. '.catch' does exactly the same as the
			second argument to '.then' does but instead you just chain it after the
			'.then' instead of adding it as a second argument. The '.catch' catches
			all errors thrown from any previousy defined '.then' above it in the
			promise chain. Promise chains continue after '.catch' so using it does not
			cancel or stop the promise from continuing.
		</p>
		<h3>Finally</h3>
		<p>
			There is a third chain method we can add to promises and thats the
			'.finally' method. This method is used at the end of a promise fo cleanup.
			Promises have 4 possible states which are PENDING, RESOLVED, REJECTED and
			SETTLED. '.then' and '.catch' cover the resolve and reject states and
			finally covers the settled state which can only ever be reached once the
			promise has finished all resolves or rejects.
		</p>
		<h3>Promise Race, Promise All, Promise All Settled</h3>
		<p>
			There are some other handy promise methods that let us deal with muiltiple
			different promises being called together. The promise race method takes an
			array of promises and returns a promise of the result of the fastest
			finishing/resolving/rejecting promise in the array of promises so you can
			then continue with chaining '.then' afterwards with that fastest promise.
			Race is useful if you are interested in working with one promise that is
			fastest of a list of promises. The promise all method allows us to work
			with a list of promises only when they have all
			finished/resolved/rejected. Here you also pass in an array of promises as
			the argument and the result you get back will be a promise containng the
			combined data of all the promises you passed into the all method. The
			combined data comes back in array form in the order of promises you passed
			into the array argument. Promise all relies on all promises you pass in to
			finish and resolve, if any one of them rejects then the whole promise all
			is cancelled and you will just get the reject error. Finally we also have
			Promise allSettled method which again takes an array of promises. The
			difference with this method is that it doesnt cancel when a promise
			fails/rejects instead it waits till all promises are settled meaning until
			all promises are either resolved or rejected and then returns that data in
			an array matching the order of the promises in the argument array. So with
			this method we get the 'all' method like behaviour but are guaranteed all
			promise data back even if some promises reject we will just get back the
			rejection error and data for that promise.
		</p>
		<h3>Async/Await</h3>
		<p>
			Async/await is a way for us to work with promises in a more synchronous
			looking way and therefore async/await is indeed still using promises under
			the hood. Async/await is used by defining an async function with the
			'async' keyword to the function statment or declaration. Adding the
			'async' keyword to the function automatically makes that function return a
			promise. With that async function defined we then get access to another
			keyword inside that function which is the 'await' keyword. We use the
			await keyword in front of any promise which then returns us the return
			data from that promise so therefore we can assign a vasriable to the
			result of await and therfore avoid the need for '.then' chaining you
			typically use with regular promises. Important to remember that
			async/await is still promises under the hood and JavaScript doesnt change
			the way they work when using async/await it is simply a way of making your
			promises more synchronous looking syntactically and JavaScript transforms
			this to work with promises and '.then' chainig etc. in the background.
		</p>
		<h3>Async/Await Error Handling</h3>
		<p>
			One issue with async/await is that we lose chaining of '.then' and
			'.catch' and therfore lose error handling with standard promises. To
			handle errors in async/await we instead use try catch blocks as the
			async/await code is more synchronous based we need a synchronous based
			error handling which is try catch.
		</p>
	</body>
</html>
